### ISSUE TITLE
Bridge: Reboot with persistant storage or Bootstrap the relayer using the blockchain

### ISSUE THREAD

**Issue by musitdev** [movementlabsxyz/movement#871]
## Is your feature request related to a problem?

Currently, when the relayer starts, it doesn't reload the state that it has when it shuts down.
If transfers are processed when the relayer stops, the transfers are lost and the relayer is not able to continue the transfer.

## Describe the solution you'd like

Since the bridge protocol is simplified, the relayer can get its state from the blockchain instead of using the indexed Db.
The bridge protocol implementation uses a nonce that can be used to order the transfer.
With the new protocol, the relayer manages only one state of transfer: pending, which exists during the time between the reception of the Initiate event and the complete transfer Tx ended.

The same algo is processed on both chain independently.
The algo has two parts:

**ongoing event processing**: this is the main relayer task where it pulls transfer event from the chains.
**reboot algo** where the relayer initiate the event polling at the same state it stops it and recover missing transfers.

### For L1-->L2 transfer scenario:

> [!NOTE]
> The algo is the same for L2-->L1 transfer, but the L1 and L2 labeling are exchanged.

**Event processing scenario :**

* get `initiate_transfer_L1L2` event in L1 block `L1block`
* save the nonce `L1nonce` and block number `L1block` in the persistent storage
* Send a `RelayEvent` and process it until the transfer is completed on L2
* Set locally the nonce `L1nonce` as processed 
* If no nonce before `L1nonce` is processed, then set the last finalized nonce `finalized_L1nonce`

**Reboot algo scenario:**

* get last finalized nonce and saved block number (`finalized_L1nonce`, `L1block`) from persistent storage.
* get current block number `current_L1block` from the chain L1.
* start the event-pulling process at the saved `L1block` block number for the chain L1
* it's the same process as normal pulling but before sending the RelayEvent, the nonce is verified (see later the nonce verification process).
* The verification is done until the process reaches the `current_L1block` block on L1.
*  Set the `last_L1nonce`, the nonce of the last process event before the `current_L1block` block.
* start the normal L1 Event processing (see above)
* start the nonce verification process.
  * Verify all nonce between `finalized_L1nonce` (exclude) and `last_L1nonce` (exclude) :
     *  query L1 contract `L1nonce->L1L2_transfer_id` map and test if the `L1L2_transfer_id` is present.
     *  if not, it means there's an issue. A `L1nonce` before the current `L1nonce` has no transfer: log error.
     *  if yes, verify on L2 that the transfer has been completed. _Rem: This data can be not finalized_.
     *  if not, query the L1 `L1L2_transfer_id` -> `details` map to get the transfer data.
     *  process the missing transfer as a normal transfer.

In this process, L1 and L2 can be rebooted in //.
This algo suppose that the contract's complete transfer function verify that a transfer hasn't been already done because the relayer can call complete_transfer more than one time for the same transfer.

This algo is compatible with using finalized blocks (see the remark on L2 complete verification). By pulling only finalized block, the `finalized_L1nonce` and `last_L1nonce` nonce are always update on finalization so
* only finalized event are processed by the normal pulling event
* only finalized `L1nonce` are verified.

## Describe alternatives you've considered
Use the indexer DB. It creates a dependency to another component. For example, if the postgres db has an issue, the relayer has an issue event if the relayer doesn't need the db to process.
We can use only the nonce and initiate the `last_L1nonce` with it. The difficulty is we are not sure that this value correspond to the current block and the event pulling use block so we can miss an transfer if we don't validate the first block we get correspond to this nonce. That why the algo use only block to update nonce.

## Additional context
This algo is defined using the new native bridge protocol with 2 functions call `Initiate_transfer` and `Complete_transfer`.

**Comment by apenzk** [comment]
Great minds think alike :D

Please checkout https://github.com/movementlabsxyz/movement/issues/872

Following issues with the above: 
> Event processing scenario will be:
> ...
> - Set the nonce N as processed. 
> - If no nonce before N is processed then set the last finalized nonce FN

This does not work as transfer completions may arrive out of order

> get current block number CB from the chain.

please specify if you refer to target or source chain across the above proposal. it is not clear to me when you talk about source chain or target chain. 

> Set the last bootstraped nonce LN

LN not defined. what is it?

> Use the indexer DB. It create a dependency to another component

it creates another trust assumption. this is not something we want to rely on with the bridge. the bridge must be safe.

**Comment by musitdev** [comment]
> This does not work as transfer completions may arrive out of order

Yes, but at one moment there's a nonce that has all previous transfer done. This is the finalized Nonce. The relayer keep track of all processed nonce that are not finalized.
For example, at one moment, the nonce 10 is finalized. The nonce 12 is processed, keep it, nonce 13 is processed keep it, nonce 11 is processed, nonce 13 become finalized. So we are sure that all transfer before this nonce are finalized.

The issue that can arrive is a transfer is never processed and need refund. This case has to be handled in a specific way because the associated transfer is always redone by the relayer so when the refund arrive and event should be sent to update the relayer state. This is not related to the bootstrap, but more how we manage failing transfer globally.

**Comment by musitdev** [comment]
In the [872](https://github.com/movementlabsxyz/movement/issues/872), from what I've understand, you describe a similar algorithm. What I try to do with this one that is not present in your algo is:
 * avoid loading all onchain transfer id for performance reason.
 * force all processed on chain data to be finalized by only request the one that correspond to a finalized block if the relayer only fetch finalized block.

Last point about depending on a local persistent data (FN, LB), they can be stored onchain but it cost, so if they are lost the relayer will restart from zero and will process all transfer that will be refused so no Tx only onchain query. It's not optimal so by using a local file you can update them by hand. It deosn't change the trust assumption because the whole relayer must be trusted in our process.

**Comment by apenzk** [comment]
I think your described process is for a relayer that restarts, not a bootstrapping of a relayer. This is a big difference.

when bootstrapping a node it **does not come with any persistant storage**, so the **only information it has is onchain**. 

> For example, at one moment, the nonce 10 is finalized. The nonce 12 is processed, keep it, nonce 13 is processed keep it, nonce 11 is processed, nonce 13 become finalized.

So actually the title of this issue should be 
"**Restarting a relayer with persistant storage**", or something like that, which differs from #872  (which describes bootstraping a fresh relayer)


> In the #872, from what I've understand, you describe a similar algorithm.

I think your algo alines with point 3) in #872 . While the on-chain bootstrapping requires something like point 4) in #872

**Comment by apenzk** [comment]
>  so if they are lost the relayer will restart from zero and will process all transfer that will be refused so no Tx only onchain query.

so i  agree. this would convert the model into a bootstrap mechanism, but it misses aspects of when and in what frequency does the relayer call what function on the contract..  also is a separate transaction required? (yes i think so) and how to ensure the relayer does not run out of gas for this action. This is explored in parts in #872

what is described in this issue is a reboot mechanism, so i took the liberty to rename the issue

**Comment by musitdev** [comment]
In fact, it's a bootstrap algo. If no persisted data are found, the algo start from zero and load all transfer id like yours. The difference is that my algo write state on disk and yours onchain. I still can store these data onchain, but less often. Once a day for example, but if you do a backup it can be the same.

In my algo I suppose that you don't have access to finalized data whereas yours. I do that because our current API doesn't allow specifying a finalized tag to fetch the data. But I found another API to get raw onchain data that allow to have only finalized data. It's more low level but doable. So I think I can simplify some part, I can get directly the on chain data without forcing the corresponding block is finalized. Not sure if it will be simpler.

I don't understand the gas issue. The algo doesn't contain any Tx. All on chain data are read only.

As I understand, the point 4) of your algo is not really feasible. You can't process a thousand of transfer id on chain and I don't see how you limit the number of verification.

I think the main different between our algo is that yours uses only onchain data to build the transfer list, where mine use block to get the event data.
I used block for 2 reasons:
 * the normal even monitoring process use block, so the same function can be used for both
 * block order the data so the switch between the bootstrap and monitoring is easier. If you get the bootstrap transfer id from the chain then you get the data from the block how you're sure not to miss an event because the on chain transfer data are not liked to a block.
 *  Initially I use it to be able to switch to finalized data too cause of the API that doesn't support finalized data.
